consensus/src/sender_aware_shuffler.rs

```diff
@@ -0,0 +1,473 @@
+    counters::{NUM_SENDERS_IN_BLOCK, TXN_SHUFFLE_SECONDS},
+    transaction_shuffler::TransactionShuffler,
+};
+use aptos_types::transaction::SignedTransaction;
+use move_core_types::account_address::AccountAddress;
+use std::{
+    cmp::min,
+    collections::{HashMap, VecDeque},
+};
+
+/// An implementation of transaction shuffler, which tries to spread transactions from same senders
+/// in a block in order to reduce conflict. On a high level, it works as follows - It defines a
+/// `conflict_window_size`, which maintains a set of senders added to the block in last `conflict_window_size`
+/// transactions. When trying to select a new transaction to the block, the shuffler tries to find
+/// a transaction which are not part of the conflicting senders in the window. If it does, it adds
+/// the first non-conflicting transaction it finds to the block, if it doesn't then it preserves the
+/// order and adds the first transaction in the remaining block. It always maintains the following
+/// invariant in terms of ordering
+/// 1. Relative ordering of all transactions from the same before and after shuffling is same
+/// 2. Relative ordering of all transactions across different senders will also be maintained if they are
+/// non-conflicting. In other words, if the input block has only one transaction per sender, the output
+/// ordering will remain unchanged.
+///
+/// In order to make the look up faster, it creates an index of contiguous set of transactions by
+/// senders. This makes searching for non-conflicting transactions much faster as compared to brute
+/// force approach.
+///
+/// Another optimization introduced is the `look_ahead_sender_window` - this limits the number of senders
+/// to check to find non-conflicting transactions before giving up. This is needed to ensure the shuffling
+/// is not O(n) even in the worst case.
+
+pub struct SenderAwareShuffler {
+    conflict_window_size: usize,
+    look_ahead_sender_window: usize,
+}
+
+impl TransactionShuffler for SenderAwareShuffler {
+    fn shuffle(&self, txns: Vec<SignedTransaction>) -> Vec<SignedTransaction> {
+        let _timer = TXN_SHUFFLE_SECONDS.start_timer();
+
+        // maintains the intermediate state of the shuffled transactions
+        let mut sliding_window = SlidingWindowState::new(self.conflict_window_size);
+        let num_transactions = txns.len();
+        let mut candidate_txn_chunks = self.prepare_txn_chunk_by_senders(txns);
+        for _ in 0..num_transactions {
+            let max_lookup = min(self.look_ahead_sender_window, candidate_txn_chunks.len());
+            // These are the chunks, which are evaluated in each iteration. These needs to be pushed
+            // back to preserve the original order of the transactions after each iteration.
+            let mut to_be_pushed_back_chunks = VecDeque::new();
+            let mut candidate_found = false;
+            let mut current_sender_index = 0;
+            while current_sender_index < max_lookup {
```

Comment:
@igor-aptos - That is a great idea but unless I am missing something, above approach will still have O(n^2) complexity in the worst case where all the remaining txns has conflict with current window and we reach to the end. 

Consider following sequence 
A1, B1, C1, A2, A3, A4, A5, B2, B3, B4, B5, C2, C3, C4, C5, 

For window size of 3, once we add A1, B1, C1 to the block, we need to scan to the end and find out that there is no non-conflicting txns. In that case, we would have to add A2 and rescan right? Or do you have anything else in mind?

Reply:
we have a pending structure: dict<sender, list<txnx>>

we read every transaction from the initial sequence once, we either put it into end sequence, or into pending.
we move every transaction once from pending final sequence.

so it is strictly O(n)

in your case (and let's say window is 10), after first 3:
sliding_window.start_index = -7
sliding_window.txns = A1, B2, C1
pending = {}

then we read the rest , putting them into pending one by one, at the end we have:
sliding_window.start_index = -7
sliding_window.txn = A1, B2, C1
pending = {A -> [A2, A3, A4, A5], B -> [B2, B3, B4, B5], C -> [C2, C3, C4, C5]}

now we get into else for " if we have more transactions to process:" where we increase sliding_window.start_index, it will go from -7 until 0, when we will see that owner of sliding_window.txn[0] is A, so we take first transaction from pending for owner A:

sliding_window.start_index = 0
sliding_window.txn = A1, B2, C1, A2
pending = {A -> [A3, A4, A5], B -> [B2, B3, B4, B5], C -> [C2, C3, C4, C5]}

then start_index++, and sliding_window.txn[1].owner is B, so we take from B:

sliding_window.start_index = 1
sliding_window.txn = A1, B2, C1, A2, B2
pending = {A -> [A3, A4, A5], B -> [B3, B4, B5], C -> [C2, C3, C4, C5]}


etc


this also removes need to have any special casing of consecutive processing of same owner (i.e. chunking)